x-pack/solutions/security/plugins/cloud_security_posture/server/routes/graph/fetch_graph.ts

```diff
@@ -172,64 +177,130 @@ const buildEsqlQuery = ({
-| MV_EXPAND target.entity.id
+| WHERE event.action IS NOT NULL AND (${actorFieldsWhereClause})
+| EVAL actorEntityId = COALESCE(
+    ${actorFieldsCoalesce}
+  )
+| EVAL targetEntityId = user.target.entity.id
+| EVAL targetEntityId = CASE(
+    host.target.entity.id IS NULL,
+    targetEntityId,
+    CASE(
+      targetEntityId IS NULL,
+      host.target.entity.id,
+      MV_DEDUPE(MV_APPEND(targetEntityId, host.target.entity.id))
+    )
+  )
+| EVAL targetEntityId = CASE(
+    service.target.entity.id IS NULL,
+    targetEntityId,
+    CASE(
+      targetEntityId IS NULL,
+      service.target.entity.id,
+      MV_DEDUPE(MV_APPEND(targetEntityId, service.target.entity.id))
+    )
+  )
+| EVAL targetEntityId = CASE(
+    entity.target.id IS NULL,
+    targetEntityId,
+    CASE(
+      targetEntityId IS NULL,
+      entity.target.id,
+      MV_DEDUPE(MV_APPEND(targetEntityId, entity.target.id))
+    )
+  )
+| MV_EXPAND actorEntityId
+| MV_EXPAND targetEntityId
+| EVAL actorEntityFieldHint = CASE(
+    user.entity.id IS NOT NULL, "user",
+    host.entity.id IS NOT NULL, "host",
+    service.entity.id IS NOT NULL, "service",
+    entity.id IS NOT NULL, "entity",
+    ""
+  )
+| EVAL targetEntityFieldHint = CASE(
+    MV_CONTAINS(user.target.entity.id, targetEntityId), "user",
+    MV_CONTAINS(host.target.entity.id, targetEntityId), "host",
+    MV_CONTAINS(service.target.entity.id, targetEntityId), "service",
+    MV_CONTAINS(entity.target.id, targetEntityId), "entity",
+    ""
+)
 ${
   isEnrichPolicyExists
     ? `
-| ENRICH ${enrichPolicyName} ON actor.entity.id WITH actorEntityName = entity.name, actorEntityType = entity.type, actorEntitySubType = entity.sub_type, actorHostIp = host.ip
-| ENRICH ${enrichPolicyName} ON target.entity.id WITH targetEntityName = entity.name, targetEntityType = entity.type, targetEntitySubType = entity.sub_type, targetHostIp = host.ip
+
+| ENRICH ${enrichPolicyName} ON actorEntityId WITH actorEntityName = entity.name, actorEntityType = entity.type, actorEntitySubType = entity.sub_type, actorHostIp = host.ip
+| ENRICH ${enrichPolicyName} ON targetEntityId WITH targetEntityName = entity.name, targetEntityType = entity.type, targetEntitySubType = entity.sub_type, targetHostIp = host.ip
 
 // Construct actor and target entities data
-| EVAL actorDocData = CONCAT("{",
-    "\\"id\\":\\"", actor.entity.id, "\\"",
-    ",\\"type\\":\\"", "${DOCUMENT_TYPE_ENTITY}", "\\"",
-    ",\\"entity\\":", "{",
-      "\\"name\\":\\"", COALESCE(actorEntityName, ""), "\\"",
-      ",\\"type\\":\\"", COALESCE(actorEntityType, ""), "\\"",
-      ",\\"sub_type\\":\\"", COALESCE(actorEntitySubType, ""), "\\"",
-      CASE (
+// Build entity field conditionally - only include if any enrichment field exists
+| EVAL actorEntityField = CASE(
+    actorEntityName IS NOT NULL OR actorEntityType IS NOT NULL OR actorEntitySubType IS NOT NULL,
+    CONCAT(",\\"entity\\":", "{",
+      "\\"name\\":\\"", COALESCE(actorEntityName, "undefined"), "\\"",
```

Comment:
here you return `undefined` as a string. Because it is wrapped with `"`.

It doesn't make much sense to me. I'd expect it to be truly `undefined`.


Also, I advise to create a helper function for generating the json in ESQL. So you won't need to repeat it everywhere.